core/src/main/java/org/apache/brooklyn/core/mgmt/persist/DeserializingClassRenamesProvider.java

```diff
@@ -20,65 +20,217 @@
+            cache = null;
+        }
+    }
+
+    public interface Loader {
+        public Map<String, String> load();
     }
     
-    private static Map<String, String> loadDeserializingClassRenamesCache() {
-        InputStream resource = new ResourceUtils(DeserializingClassRenamesProvider.class).getResourceFromUrl(DESERIALIZING_CLASS_RENAMES_PROPERTIES_PATH);
-        if (resource != null) {
+    /**
+     * Loads the class-renames from the OSGi configuration file: {@code org.apache.brooklyn.classrename.cfg}.
+     * 
+     * Only public for OSGi instantiation - treat as an internal class, which may change in 
+     * future releases.
+     * 
+     * See http://stackoverflow.com/questions/18844987/creating-a-blueprint-bean-from-an-inner-class:
+     * we unfortunately need to include {@code !org.apache.brooklyn.core.mgmt.persist.DeserializingClassRenamesProvider}
+     * in the Import-Package, as the mvn plugin gets confused due to the use of this inner class
+     * within the blueprint.xml.
+     * 
+     * @see {@link DeserializingClassRenamesProvider#KARAF_DESERIALIZING_CLASS_RENAMES_PROPERTIES}
+     */
+    public static class OsgiConfigLoader implements Loader {
+        private ConfigurationAdmin configAdmin;
+
+        public OsgiConfigLoader() {
+            LOG.trace("OsgiConfigLoader instance created");
+        }
+
+        // For injection as OSGi bean
+        public void setConfigAdmin(ConfigurationAdmin configAdmin) {
+            this.configAdmin = configAdmin;
+        }
+
+        // Called by OSGi
+        public void init() {
+            LOG.trace("DeserializingClassRenamesProvider.OsgiConfigLoader.init: registering loader");
+            DeserializingClassRenamesProvider.loaders.add(this);
+            DeserializingClassRenamesProvider.reset();
+        }
+
+        // Called by OSGi
+        public void destroy() {
+            LOG.trace("DeserializingClassRenamesProvider.OsgiConfigLoader.destroy: unregistering loader");
+            boolean removed = DeserializingClassRenamesProvider.loaders.remove(this);
+            if (removed) {
+                DeserializingClassRenamesProvider.reset();
+            }
+        }
+
+        // Called by OSGi when configuration changes
+        public void updateProperties(Map properties) {
+            LOG.debug("DeserializingClassRenamesProvider.OsgiConfigLoader.updateProperties: clearing cache, so class-renames will be reloaded");
+            DeserializingClassRenamesProvider.reset();
+        }
+
+        @Override
+        public Map<String, String> load() {
+            if (configAdmin == null) {
+                LOG.warn("No OSGi configuration-admin available - cannot load {}.cfg", KARAF_DESERIALIZING_CLASS_RENAMES_PROPERTIES);
+                return ImmutableMap.of();
+            }
+            
+            String filter = '(' + Constants.SERVICE_PID + '=' + KARAF_DESERIALIZING_CLASS_RENAMES_PROPERTIES + ')';
```

Comment:
Add a TODO here, you could avoid using ConfigurationAdmin in the code and doing all that filtering/ listConfigurations yourself. It should be possible to get cm to do that for you with a managed bean;  but  it’s not a big deal, it just means you have written a little more code  than you needed to.

The way the configuration admin service is typically intended to be used is that you let it inject properties into the beans that you declare - i.e. your code doesn't need to get hold of the ConfigurationService class at all.  See https://www.eclipse.org/gemini/blueprint/documentation/reference/2.0.0.RELEASE/html/compendium.html section 11.1.2.1 for background.

I was thinking along the lines of changing the DeserializingClassRenamesProvider to be a regular bean that could be injected as a service, but the pattern of callers would make that a quite extensive change, affecting a lot of classes.  But it shouldn't be hard to make it a singleton bean, and have it be initialised by the launchers. Then in Karaf (OsgiLauncher) you would have an "DeserializingClassRenamesProviderInitializer" managed bean in the blueprint.xml,

```xml
<bean id="deserializingRenamesInit" class="org.apache.brooklyn.launcher.osgi.DeserializingClassRenamesProviderInitializer">
   <osgix:managed-properties persistent-id="org.apache.brooklyn.core.deserializing.renames" update-method="updateCallback"/>
</bean>
```

and then the DeserializingClassRenamesProviderInitializer would have a method to initialize the provider, along the following lines and ignoring all niceties of synchronization, exceptions etc.

```java
public class DeserializingClassRenamesProviderInitializer {

    public void updateCallback(Map properties) {

        DeserializingClassRenamesProvider.initializeSingleton(properties);
    }
}
```

and changing the provider class to

```java
public class DeserializingClassRenamesProvider {

    private volatile Map<String, String> cache;

    public static DeserializingClassRenamesProvider INSTANCE;

    private DeserializingClassRenamesProvider(Map properties) {
        // reset cache from properties
    }

    public static void initializeSingleton(Map properties) {
        INSTANCE = new DeserializingClassRenamesProvider(properties);
    }

    // not static!
    public Map<String, String> loadDeserializingClassRenames() {
       // return the cache
    }
}
```